lib/backends/smartos/lib/docker-stdio.js

```diff
@@ -449,19 +473,28 @@ function createDockerFileReadStreamServer(opts, callback) {
                     log.error('onConnectionZoneRunning(read): ' +
                             'data timeout; terminating archive process');
                     streamProc.kill('SIGKILL');
-                }
-);
-                streamProc.stdout.pipe(socket);
+                });
+
+                // Pipe the file tar stream, but don't end the socket, as we
+                // want to properly unmount the zone before we end - that way
+                // any other calls to docker copy will wait until the zone is
+                // properly unmounted before performing a new copy.
+                streamProc.stdout.pipe(socket, {end: false});
 
                 streamProc.stdout.on('end', function _onReadEnd() {
-                    next();
+                    tryUnmountLocal({ uuid: uuid, log: log },
+                            function _unmountCb(err) {
+                        log.info('zone unmounted - closing the socket now');
+                        socket.end();
+                        next(err);
+                    });
```

Comment:
I really don't see the need for a callback function only _some times_. I'd say that if the unmount operation is blocking the process somehow, the callback should be present always. Otherwise, I'd just call everything within the callbacks right after calling the func and forget about the callback completely.

Reply:
Not sure what you mean by only sometimes?

Reply:
If you check the `tryUnmountLocal` function, the callback is optional. It's used w/o callback at line 498, for example.